xbmc/cores/AudioEngine/Sinks/AESinkStarfish.cpp

```diff
@@ -0,0 +1,350 @@
+                  m_format.m_channelLayout.Count());
+        return false;
+    }
+    payload["option"]["externalStreamingInfo"]["contents"]["pcmInfo"]["channelMode"] = channel;
+    auto pcmFormat = AEFormatToStarfishFormat(m_format.m_dataFormat);
+    if (pcmFormat.empty())
+    {
+      CLog::Log(LOGWARNING, "CAESinkStarfish::Initialize PCM format is empty");
+      return false;
+    }
+    payload["option"]["externalStreamingInfo"]["contents"]["pcmInfo"]["format"] = pcmFormat.data();
+    payload["option"]["externalStreamingInfo"]["contents"]["codec"]["audio"] = "PCM";
+  }
+
+  payload["option"]["externalStreamingInfo"]["bufferingCtrInfo"]["preBufferByte"] = 0;
+  payload["option"]["externalStreamingInfo"]["bufferingCtrInfo"]["bufferMinLevel"] = 0;
+  payload["option"]["externalStreamingInfo"]["bufferingCtrInfo"]["bufferMaxLevel"] = 0;
+  // This is the size after which the sink starts blocking
+  payload["option"]["externalStreamingInfo"]["bufferingCtrInfo"]["qBufferLevelAudio"] =
+      m_bufferSize;
+  // Internal buffer?
+  payload["option"]["externalStreamingInfo"]["bufferingCtrInfo"]["srcBufferLevelAudio"]["minimum"] =
+      format.m_frames;
+  payload["option"]["externalStreamingInfo"]["bufferingCtrInfo"]["srcBufferLevelAudio"]["maximum"] =
+      m_bufferSize;
+
+  CVariant payloadArgs;
+  payloadArgs["args"] = CVariant(CVariant::VariantTypeArray);
+  payloadArgs["args"].push_back(std::move(payload));
+
+  std::string json;
+  CJSONVariantWriter::Write(payloadArgs, json, true);
+
+  m_starfishMediaAPI->notifyForeground();
+  CLog::Log(LOGDEBUG, "CAESinkStarfish: Sending Load payload {}", json);
+  if (!m_starfishMediaAPI->Load(json.c_str(), &CAESinkStarfish::PlayerCallback, this))
+  {
+    CLog::Log(LOGERROR, "CAESinkStarfish::Initialize Load failed");
+    return false;
+  }
+
+  return true;
+}
+
+void CAESinkStarfish::Deinitialize()
+{
+  m_starfishMediaAPI->Unload();
+}
+
+double CAESinkStarfish::GetCacheTotal()
+{
+  if (m_format.m_dataFormat == AE_FMT_RAW)
+    return 8 * m_format.m_streamInfo.GetDuration();
+  else
+    return 0.0;
+}
+
+double CAESinkStarfish::GetLatency()
+{
+  return 0.0;
+}
+
+unsigned int CAESinkStarfish::AddPackets(uint8_t** data, unsigned int frames, unsigned int offset)
+{
+  CVariant payload;
+  auto buffer = data[0] + offset * m_format.m_frameSize;
+  payload["bufferAddr"] = fmt::format("{:#x}", reinterpret_cast<std::uintptr_t>(buffer));
+  payload["bufferSize"] = frames * m_format.m_frameSize;
+  payload["pts"] = m_pts;
+  payload["esData"] = 2;
+
+  int64_t frameTime;
+  if (m_format.m_dataFormat == AE_FMT_RAW)
+    frameTime = static_cast<int64_t>(m_format.m_streamInfo.GetDuration() * 1000000.0);
+  else
+    frameTime = 1000000000 * frames / m_format.m_sampleRate;
+
+  m_pts += frameTime;
+
+  std::string json;
+  CJSONVariantWriter::Write(payload, json, true);
+
+  auto result = m_starfishMediaAPI->Feed(json.c_str());
+  while (result.find("BufferFull") != std::string::npos)
```

Comment:
don't want to hold this up any more, so for a future PR: this is prone to an endless loop if the API never gives you BufferFull for whatever reason -> needs hard iteration limit